### ISSUE TITLE
syntactic_sugar incorrectly warns about syntactic sugar on arrays when accessing associated type

### ISSUE THREAD

**Issue by mdznr** [realm/SwiftLint#3502]
When using `syntactic_sugar`, it appears to incorrectly warn about something like:

```swift
extension Sequence where Element: Collection {
  func foo() -> [Array<Element.Element>.Index] {
    var indices = [Array<Element.Element>.Index]() // ⚠️ Syntactic Sugar Violation: Shorthand syntactic sugar should be used, i.e. [Int] instead of Array<Int>. (syntactic_sugar)
    return indices
  }
}
```

However, Swift (5.3.2) won’t let you write what SwiftLint is suggesting in this case:

```swift
extension Sequence where Element: Collection {
  func foo() -> [Array<Element.Element>.Index] {
    var indices = [[Element.Element].Index]() // 🛑 Cannot call value of non-function type '[Array<Self.Element.Element>.Index.Type]' (aka 'Array<Int.Type>')
    return indices
  }
}
```

----
SwiftLint v0.34.0

**Comment by personalnadir** [comment]
I think I'm seeing the same issue when accessing the Element typealias in dictionaries - I get the syntactic sugar warning, but using it causes a compile error. However it only occurs in the context of a closure for example:

```
dictionary.reduce(0, {(count: Int, element: Dictionary<String, Any>.Element) -> Int in
            if let subdictionary = element.value as? [String: Any] {
                return subdictionary["field"] != nil ? count + 1 : count
            }

            return count
        })
```

**Comment by SimplyDanny** [comment]
@PaulTaykalo, this might be something for you. 😉
